### PR TITLE
Deregister from ALBs

### LINKED ISSUE #2: ALB Deregister: post drain, should deregister instance from all available target groups
In order to make ALB draining better with `alb-ingress-controller`, we should allow `--with-deregister` flag, which would go to all target groups found and start deregistering the instance from them, and then wait.

This can make 5xx errors much less frequent.

We can leverage following label:
`alpha.service-controller.kubernetes.io/exclude-balancer: true`

https://github.com/kubernetes-sigs/aws-alb-ingress-controller/issues/325
https://github.com/kubernetes-sigs/aws-alb-ingress-controller/pull/403/files

### PR DESCRIPTION
**Pull request by eytan-avisror**
This PR fixes #2 and adds the following features:

- use `--with-deregister` flag is on by default
- following label `alpha.service-controller.kubernetes.io/exclude-balancer: true` will be added to node.
- controller will scan all target groups and attempt to find a target matching with the instance-id, and run a deregister request

This completely removes 5xx errors on scale down, here are some tests performed to verify this.

### Test Scenario
Two node groups, `foo` and `bar`, node group `foo` has an nginx application running with an ingress object, node group `bar` has 10 nodes running but no apps.

Traffic is sent via `locust` about 150TPS to the ingress object.

During load, `bar` node group gets scaled up and down every few minutes (+/- 5 nodes).

### Results
Blue - request count (x axis)
Red - 5xx errors (y axis)
Orange - Unhealthy targets (y axis)

#### with lifecycle-manager disabled
Screen Shot 2019-10-03 at 9 21 58 PM

Total Requests: 482941 (+/- 150TPS x 45minutes)
Failed 5xx: 137 (0.028%)

#### with lifecycle-manager enabled + `--with-deregister` flag set to `false`
Screen Shot 2019-10-06 at 2 54 49 PM

Total Requests: 475512 (+/- 150TPS x 45minutes)
Failed 5xx: 132 (0.03%)

#### with lifecycle-manager enabled + `--with-deregister` flag set to `true`
Screen Shot 2019-10-05 at 10 26 08 AM

Total Requests: 482823 (+/- 150TPS x 45minutes)
Failed 5xx: 0

### CODE REVIEW

**Comment by eytan-avisror** [comment]
#### Second `--with-deregister` test:
Screen Shot 2019-10-06 at 6 14 32 PM

Total Requests: 500970 (+/- 150TPS x 45minutes)
Failed 5xx: 1

**Comment by shrinandj** [comment]
Amazing feature, great tests and results :-)

Few questions though:

> controller will scan all target groups and attempt to find a target matching with the instance-id, and run a deregister request

Why does the controller need to scan target-groups? The lifecycle-manager can simply add the label to the node irrespective of whether it is in any target group or not, right?

Also, instead of having ingress specific labels/annotations, can we have generic labels/annotations which get applied to every node for which the lifecycle-manager gets requests?

**Comment by eytan-avisror** [comment]
> Amazing feature, great tests and results :-)
> 
> Few questions though:
> 
> > controller will scan all target groups and attempt to find a target matching with the instance-id, and run a deregister request
> 
> Why does the controller need to scan target-groups? The lifecycle-manager can simply add the label to the node irrespective of whether it is in any target group or not, right?
> 
> Also, instead of having ingress specific labels/annotations, can we have generic labels/annotations which get applied to every node for which the lifecycle-manager gets requests?

Thanks! to answer your questions:
- Unfortunately, simply adding the label `alpha.service-controller.kubernetes.io/exclude-balancer: true` does not immediately trigger a draining, it just means that the next time the controller scans for targets/nodes, it will not see the node as a target to be added, so we must actively trigger the deregister from all the target groups the node is a member of, if you continue with termination before instance is fully drained, you will get 5xx errors.

- Regarding generic - I thought about this quite a bit, and figured the label itself is quite generic, not sure about other ingress controllers and whether they use the same mechanism, but it did really make sense to me to make this 'any' label, because this was already specific to `alb-ingress-controller` by virtue of deregistering from target groups on AWS.

**Comment by eytan-avisror** [comment]
@kevdowney the label is just from excluding the instance from load balancers (e.g. don't try to register it on reconciles). This controller then goes ahead and trigger the deregister from target group after adding this label.
A taint won't help us here as it's not a scheduling issue.

File: pkg/service/types.go [hunk 1]
```diff
@@ -4,6 +4,8 @@ import (
 	"reflect"
 	"sync"
 
+	"github.com/aws/aws-sdk-go/service/elbv2/elbv2iface"
+
```
Comment (davemasselink) [inline pkg/service/types.go]:
```suggestion
```

File: pkg/service/elbv2_test.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
+package service
+
+import (
+	"context"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws/request"
+
```
[the remark this answers is not in the document]
Comment (eytan-avisror) [reply pkg/service/elbv2_test.go]: This is actually recommended in go in order to separate local dependencies and external ones.
I have a plugins that does this by default

File: pkg/service/elbv2.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+	"github.com/aws/aws-sdk-go/service/elbv2/elbv2iface"
+
+	"github.com/keikoproj/lifecycle-manager/pkg/log"
+)
+
+func waitForDeregisterInstance(elbClient elbv2iface.ELBV2API, arn, instanceID string, port int64) error {
+	var (
```
[the remark this answers is not in the document]
Comment (eytan-avisror) [reply pkg/service/elbv2.go]: const may make things harder in testing, there is no real reason to use it outside of Enums or few other use cases

**Pull request merged**